### PR TITLE
Event handlers forwarded to a window check for window presence

### LINKED ISSUE #15332: Event handlers forwarded to a window should check for the presence of a window
From https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-idl-attributes:
```
If an event handler IDL attribute exposes an event handler of an object that doesn't exist, it must always return null on getting and must do nothing on setting.

This can happen in particular for event handler IDL attribute on body elements that do not have corresponding Window objects.
```
Our implementation of `define_window_owned_event_handler` does not check whether it's part of an element that is in a document that has a browsing context, and it should. This JS behaves differently in Servo vs. other browsers:
```html
<script>
var d = (new DOMParser).parseFromString('', 'text/html');
var b = d.createElement('body');
b.onload = function() {};
console.log(b.onload);
</script>
```
In Servo it logs `function() {}`, in other browsers it logs `null`.

Code: `components/script/dom/macros.rs`
Test: We should [create a test](https://github.com/servo/servo/blob/master/tests/wpt/README.md#writing-new-tests) in `tests/wpt/web-platform-tests/html/webappapis/scripting/events/` that verifies this behaviour.

### PR DESCRIPTION
**Pull request by freeatnet**
This change adds a check for the presence of a window in event handlers of `body` and `frameset` that are forwarded to the window.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15332

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

I'm new to both Servo and Rust — so any suggestions for improvement are very welcome.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15359)

### CODE REVIEW

**Comment by jdm** [comment]
Great work on the tests! All the changes look correct; please go ahead and do the non-native handlers too!

File: components/script/dom/htmlelement.rs [hunk 1]
```diff
@@ -151,7 +151,11 @@ impl HTMLElementMethods for HTMLElement {
     // https://html.spec.whatwg.org/multipage/#handler-onload
     fn GetOnload(&self) -> Option<Rc<EventHandlerNonNull>> {
         if self.is_body_or_frameset() {
-            window_from_node(self).GetOnload()
+            if document_from_node(self).has_browsing_context() {
+                window_from_node(self).GetOnload()
+            } else {
+                None
```
Comment (KiChjang) [inline components/script/dom/htmlelement.rs]: Why is this returning `None` here? Shouldn't it default into returning `self.upcast::<EventTarget>().get_event_handler_common("load")`?

File: components/script/dom/htmlelement.rs [hunk 2]
```diff
@@ -169,7 +175,11 @@ impl HTMLElementMethods for HTMLElement {
     // https://html.spec.whatwg.org/multipage/#handler-onresize
     fn GetOnresize(&self) -> Option<Rc<EventHandlerNonNull>> {
         if self.is_body_or_frameset() {
-            window_from_node(self).GetOnload()
+            if document_from_node(self).has_browsing_context() {
+                window_from_node(self).GetOnload()
+            } else {
+                None
```
Comment (KiChjang) [inline components/script/dom/htmlelement.rs]: Same question here.

File: components/script/dom/htmlelement.rs [hunk 3]
```diff
@@ -187,7 +199,11 @@ impl HTMLElementMethods for HTMLElement {
     // https://html.spec.whatwg.org/multipage/#handler-onblur
     fn GetOnblur(&self) -> Option<Rc<EventHandlerNonNull>> {
         if self.is_body_or_frameset() {
-            window_from_node(self).GetOnblur()
+            if document_from_node(self).has_browsing_context() {
+                window_from_node(self).GetOnblur()
+            } else {
+                None
```
Comment (KiChjang) [inline components/script/dom/htmlelement.rs]: Same question here.

File: components/script/dom/htmlelement.rs [hunk 4]
```diff
@@ -205,7 +223,11 @@ impl HTMLElementMethods for HTMLElement {
     // https://html.spec.whatwg.org/multipage/#handler-onfocus
     fn GetOnfocus(&self) -> Option<Rc<EventHandlerNonNull>> {
         if self.is_body_or_frameset() {
-            window_from_node(self).GetOnfocus()
+            if document_from_node(self).has_browsing_context() {
+                window_from_node(self).GetOnfocus()
+            } else {
+                None
```
Comment (KiChjang) [inline components/script/dom/htmlelement.rs]: Same question here.

File: components/script/dom/htmlelement.rs (hunk 2 above)
Comment (KiChjang) [inline components/script/dom/htmlelement.rs]: Same question here.

File: components/script/dom/htmlelement.rs [hunk 5]
```diff
@@ -223,7 +247,11 @@ impl HTMLElementMethods for HTMLElement {
     // https://html.spec.whatwg.org/multipage/#handler-onscroll
     fn GetOnscroll(&self) -> Option<Rc<EventHandlerNonNull>> {
         if self.is_body_or_frameset() {
-            window_from_node(self).GetOnscroll()
+            if document_from_node(self).has_browsing_context() {
+                window_from_node(self).GetOnscroll()
+            } else {
+                None
```
Comment (KiChjang) [inline components/script/dom/htmlelement.rs]: Same question here.

File: components/script/dom/macros.rs [hunk 6]
```diff
@@ -354,11 +354,17 @@ macro_rules! define_event_handler(
 macro_rules! define_window_owned_event_handler(
     ($handler: ident, $event_type: ident, $getter: ident, $setter: ident) => (
         fn $getter(&self) -> Option<::std::rc::Rc<$handler>> {
-            window_from_node(self).$getter()
+            if document_from_node(self).has_browsing_context() {
+                window_from_node(self).$getter()
+            } else {
+                None
```
Comment (KiChjang) [inline components/script/dom/macros.rs]: Same question here.

File: components/script/dom/htmlelement.rs [hunk 7]
```diff
@@ -151,7 +151,11 @@ impl HTMLElementMethods for HTMLElement {
     // https://html.spec.whatwg.org/multipage/#handler-onload
     fn GetOnload(&self) -> Option<Rc<EventHandlerNonNull>> {
         if self.is_body_or_frameset() {
-            window_from_node(self).GetOnload()
+            if document_from_node(self).has_browsing_context() {
+                window_from_node(self).GetOnload()
```
Comment (KiChjang) [inline components/script/dom/htmlelement.rs]: This pattern can be simplified to
```rust
let document = document_from_node(self);
if document.has_browsing_context() {
    document.window().GetOnload()
}
```

**Review by KiChjang — CHANGES_REQUESTED** [review]

File: components/script/dom/macros.rs [hunk 8]
```diff
@@ -354,11 +354,17 @@ macro_rules! define_event_handler(
 macro_rules! define_window_owned_event_handler(
     ($handler: ident, $event_type: ident, $getter: ident, $setter: ident) => (
         fn $getter(&self) -> Option<::std::rc::Rc<$handler>> {
-            window_from_node(self).$getter()
+            if document_from_node(self).has_browsing_context() {
+                window_from_node(self).$getter()
```
Comment (freeatnet) [inline components/script/dom/macros.rs]: @KiChjang Any objections to applying the simplified pattern from above here too?

File: components/script/dom/htmlelement.rs (hunk 1 above)
Comment (freeatnet) [inline components/script/dom/htmlelement.rs]: I believe this is consistent with the following specification definition (mentioned in #15332):
> If an event handler IDL attribute exposes an event handler of an object that doesn't exist, it must always return null on getting and must do nothing on setting.

File: components/script/dom/macros.rs (hunk 8 above)
Comment (KiChjang) [inline components/script/dom/macros.rs]: No, go right ahead in fixing it!

**Comment by KiChjang** [comment]
Also, my comments from my last review hasn't been addressed.

File: components/script/dom/htmlelement.rs [hunk 9]
```diff
@@ -151,7 +151,12 @@ impl HTMLElementMethods for HTMLElement {
     // https://html.spec.whatwg.org/multipage/#handler-onload
     fn GetOnload(&self) -> Option<Rc<EventHandlerNonNull>> {
         if self.is_body_or_frameset() {
-            window_from_node(self).GetOnload()
+            let document = document_from_node(self);
+            if document.has_browsing_context() {
+                document.window().GetOnload()
+            } else {
+                None
```
Comment (KiChjang) [inline components/script/dom/htmlelement.rs]: > If an event handler IDL attribute exposes an event handler of an object that doesn't exist, it must always return null on getting and must do nothing on setting.

Here, you did not check whether an event handler exists when a browsing context doesn't exist.

Comment (freeatnet) [inline components/script/dom/htmlelement.rs]: Is there a case where a window would exist without a browsing context?

Comment (KiChjang) [inline components/script/dom/htmlelement.rs]: That's orthogonal to the problem I described. If there isn't a browsing context (i.e. there's not a window), then you should check whether the HTML element has an event handler registered for this particular event.

Comment (jdm) [inline components/script/dom/htmlelement.rs]: @KiChjang That countermands the spec, as @freeatnet pointed out.

Comment (freeatnet) [inline components/script/dom/htmlelement.rs]: The branch we are discussing deals with `body` and `frameset` elements that do not have a browsing context only, right?

My understanding is that `body` and `frameset` elements actually expose event handlers of their Window as their event handlers (first note in [the spec](https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-idl-attributes)).

Is there a case where the note above does not hold?

**Comment by KiChjang** [comment]
Okay, nevermind, confusion has been solved via IRC chat with @jdm.

@bors-servo r+

**Comment by freeatnet** [comment]
@KiChjang Thank you and my apologies for the confusion. I should've linked the related specs in the PR.

**Comment by jgraham** [comment]
Please note that linking to external resources (in this case jQuery) in tests isn't allowed because it is commonly forbidden in tests environments and where it isn't may lead to test instability.

**Comment by freeatnet** [comment]
@jgraham oh shoot. It was accidentally left in and isn't actually required for the test. Want me to PR the removal?

**Comment by jgraham** [comment]
I'm going to land a commit in Gecko to remove it, which will be upstreamed in due course. But a separate PR here or in web-platfrom-tests wouldn't be harmful.

**Pull request merged**